---
title: "Assignment: Spatial Diversity"
author: "Katie Beidler; Z620: Quantitative Biodiversity, Indiana University"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: pdf_document
geometry: margin=2.54cm
---

## OVERVIEW
This assignment will emphasize primary concepts and patterns associated with spatial diversity, while using R as a Geographic Information Systems (GIS) environment. 
Complete the assignment by refering to examples in the handout.

After completing this assignment you will be able to:  
1.  Begin using R as a geographical information systems (GIS) environment.  
2.  Identify primary concepts and patterns of spatial diversity.  
3.  Examine effects of geographic distance on community similarity.  
4.  Generate simulated spatial data.

## Directions:
1. Change "Student Name" on line 3 (above) with your name.
2. Complete as much of the assignment as possible during class; what you do not complete in class will need to be done on your own outside of class.
3. Use the handout as a guide; it contains a more complete description of data sets along with the proper scripting needed to carry out the assignment.
4. Be sure to **answer the questions** in this assignment document.
Space for your answer is provided in this document and indicated by the ">" character.
If you need a second paragraph be sure to start the first line with ">".
5. Before you leave the classroom, **push** this file to your GitHub repo.
6. When you are done wit the assignment, **Knit** the text and code into an html file.
7. After Knitting, please submit the completed assignment by creating a **pull request** via GitHub.
Your pull request should include this file *spatial_assignment.Rmd* and the html output of `Knitr` (*spatial_assignment.html*).


## 1) R SETUP
In the R code chunk below, provide the code to:  

1. Clear your R environment
2. Print your current working directory,
3. Set your working directory to your "*/Week4-Spatial*" folder, and

```{r, results = 'hide'}
rm(list=ls())
getwd()
setwd("/Users/kbeidler/GitHub/QB2017_Beidler/Week4-Spatial")
```

## 2) LOADING R PACKAGES

In the R code chunk below, do the following:  

1. Install and/or load the following packages: `vegan`, `sp`, `gstat`, `raster`, `RgoogleMaps`, `maptools`, `rgdal`, `simba`, `gplots`, `rgeos`

```{r, results = 'hide', message = FALSE, warning = FALSE} 
require(vegan)
#install.packages('sp') # Classes and methods for handling spatial data require(sp)
#install.packages('gstat') # Methods for geostatistical analyses
require(gstat)
#install.packages('raster') # Methods to create a RasterLayer object require(raster)
#install.packages('RgoogleMaps') # For querying the Google server for static maps. require(RgoogleMaps)
#install.packages('maptools') # Tools for manipulating and reading geospatial data require(maptools)
#install.packages('rgdal') # Geospatial Data Abstraction Library
require(rgdal)
#install.packages('simba')
require("simba")
#install.packages('gplots')
require("gplots")
```

***Question 1***:  What are the packages `simba`, `sp`, and `rgdal` used for?

> ***Answer 1***:  The Simba package uses a collection of functions for similarity analysis of vegetation data, as well as, functions for reshaping species lists into matrices. The sp function creates classes for spatial data and provides utility functions, e.g. for plotting data as maps, spatial selection, as well as methods for retrieving coordinates, for subsetting, print, summary, etc. Rgdal is a package that helps you read in vector-based spatial data.


## 3) LOADING DATA
In the R code chunk below, use the example in the handout to do the following:  

1. Load the Site-by-Species matrix for the Indiana ponds datasets: BrownCoData/SiteBySpecies.csv
2. Load the Environmental data matrix: BrownCoData/20130801_PondDataMod.csv
3. Assign the operational taxonomic units (OTUs) to a variable 'otu.names'
4. Remove the first column (i.e., site names) from the OTU matrix.

```{r, results = 'hide', message = FALSE, warning = FALSE} 
Ponds = read.table(file = "BrownCoData/20130801_PondDataMod.csv", head = TRUE, sep = ",") 
lats = as.numeric(Ponds[, 3]) 
# latitudes (north and south)
lons = as.numeric(Ponds[, 4]) 
# longitudes (east and west)
OTUs = read.csv(file = "BrownCoData/SiteBySpecies.csv", head = TRUE, sep = ",")
# remove first column (site names)
OTUs = as.data.frame(OTUs[-1]) 

# To get S, we sum the number of non-zero abundances for a site
S.obs = function(x = ""){ rowSums(x > 0) * 1} 
otu.names = names(OTUs) # Get the names of the OTUs
Ponds$N = as.vector(rowSums(OTUs)) # numbers of reads 

Ponds$S = S.obs(OTUs) # number of species OTUs
max(Ponds$S)
Ponds$H = as.vector(diversity(OTUs, index = "shannon"))
# To get De at each site, we divide Simpson's Diversity by OTU site richness
Ponds$D = as.vector(diversity(OTUs, index = "invsimpson")/Ponds$S)
```

***Question 2a***:  How many sites and OTUs are in the SiteBySpecies matrix? 

> ***Answer 2a***:  There are 51 sites and 16,383 OTUs. 

***Question 2b***:  What is the greatest species richness found among sites?

> ***Answer 2b***:  3,659 is the greatest species richness found at site 47.


## 4) GENERATE MAPS

In the R code chunk below, do the following:  

1. Using the example in the handout, visualize the spatial distribution of our samples with a basic map in RStudio using the `GetMap` function in the package `RgoogleMaps`. 
This map will be centered on Brown County, Indiana (39.1 latitude, -86.3 longitude).

```{r}
require('RgoogleMaps')
newmap = GetMap(center = c(39.1,-86.3), zoom = 10,
destfile = "PondsMap.png", maptype="terrain")
PlotOnStaticMap(newmap, zoom = 10, cex = 2, col = 'blue') 
```

***Question 3***:  Briefly describe the geographical layout of our sites.

> ***Answer 3***:  The sites appear to be clustered around bodies of water in areas with low tree cover. 


In the R code chunk below, do the following:  

1. Using the example in the handout, build a map by combining lat-long data from our ponds with land cover data and data on the locations and shapes of surrounding water bodies.

```{r, results = 'hide', message = FALSE, warning = FALSE}
# 1. Import TreeCover.tif as a raster file.
require('raster')
Tree.Cover = raster("TreeCover/TreeCover.tif") # import TreeCover.tif as a raster file.
# 2. Plot the % tree cover data
plot(Tree.Cover, xlab = 'Longitude', ylab = 'Latitude', 
     main = 'Map of geospatial data for % tree cover,\nwater bodies, and sample sites')
# 3. Import water bodies as a shapefile.
require('maptools')
Water.Bodies = readShapeSpatial("water/water.shp") # import water bodies as a shapefile
# 4. Plot the water bodies around our study area, i.e., Monroe County
plot(Water.Bodies, border='cyan', axes = TRUE, add = TRUE)
# 5. Plot the refuge pond locations
Refuge.Ponds = SpatialPoints(cbind(lons, lats)) 
# 6. Convert lat-long data for ponds to georeferenced points.
plot(Refuge.Ponds, line='r', col='red', pch = 20, cex = 1.5, add = TRUE)
```

***Question 4a***: What are datums and projections?

> ***Answer 4a***: A datum is a model for Earthâ€™s shape and a projection is the way in which coordinates on a sphere are projected onto a 2-D surface and the distoriation that accompanies that proccess. Datums and projections are fundamental to making maps and geospatial analysis.

## 5) UNDERSTANDING SPATIAL AUTOCORRELATION

***Question 5***: In your own words, explain the concept of spatial autocorrelation.

> ***Answer 5***: Spatial autocorrelation is the idea that samples or individuals that are closer in space are more similar to one another.[Spatial autocorrelation is the degree to which spatial variables are either clustered in space or overdispersed.] Increasing similarity with decreasing geographic distance. Things that are closer in physical space are more similar to one another.

## 6) EXAMINING DISTANCE-DECAY

***Question 6***: In your own words, explain what a distance decay pattern is and what it reveals.

> ***Answer 6***: The distance-decay relationship shows the pattern of autocorrelation, how Bray Curtis dissimilarity values change as you move greater distances. Bray-curtis decreases as geographic distance increases. If there is no distance decay relationship sites are not spatially autocorrelated and there is no distance at which you can't compare species.

In the R code chunk below, do the following:  

1. Generate the distance decay relationship for bacterial communities of our refuge ponds and for some of the environmental variables that were measured. 
Note: You will need to use some of the data transformations within the *semivariogram* section of the handout.

```{r, results = 'hide', message = FALSE, warning = FALSE}
# Construct a new dataframe for coordinates
xy = data.frame(env = Ponds$TDS, pond.name = Ponds$Sample_ID, lats = Ponds$lat, lons = Ponds$long)
coordinates(xy) = ~lats+lons # Transform 'xy' into a spatial points dataframe
# Identify the current projection (i.e., lat-long) and datum (NAD83)
proj4string(xy) = CRS("+proj=longlat +datum=NAD83") # coordinate reference system (CRS)
UTM = spTransform(xy, CRS("+proj=utm +zone=51 +ellps=WGS84"))
UTM = as.data.frame(UTM)
xy$lats_utm = UTM[,2] # lattitude data according to UTM
xy$lons_utm = UTM[,3] # longitude data according to UTM

# 1) Calculate Bray-Curtis similarity between plots using the `vegdist()` function
comm.dist = 1 - vegdist(OTUs) # Bray-Curtis similarity between the plots
# 2) Assign UTM lattitude and longitude data to 'lats' and 'lons' variables
lats = as.numeric(xy$lats_utm) # lattitude data
lons = as.numeric(xy$lons_utm) # longitude data
# 3) Calculate geographic distance between plots and assign to the variable 'coord.dist'
coord.dist = dist(as.matrix(lats, lons)) # geographical distance between plots
# 4) Transform environmental data to numeric type, and assign to variable 'x1'
x1 = as.numeric(Ponds$"SpC")
# 5) Using the `vegdist()` function in `simba`, calculate the Euclidean distance between the plots for environmental variables. Assign the result to the variable 'env.dist'
env.dist = vegdist(x1, "euclidean") 
# 6) Transform all distance matrices into database format using the `liste()` function in `simba`:
comm.dist.ls = liste(comm.dist, entry="comm") 
env.dist.ls = liste(env.dist, entry="env") 
coord.dist.ls = liste(coord.dist, entry="dist")
# 7) Create a data frame containing similarity of the environment and similarity of community.
df = data.frame(coord.dist.ls, env.dist.ls[,3], comm.dist.ls[,3])
# 8) Attach the columns labels 'env' and 'struc' to the dataframe you just made.
names(df)[4:5] = c("env", "struc")
attach(df)
# 9) After setting the plot parameters, plot the distance-decay relationships, with regression lines in red.
par(mfrow=c(1, 2), pty="s")
plot(env, struc, xlab="Environmental Distance", ylab="1 - Bray-Curtis",
     main = "Environment", col='SteelBlue')

OLS = lm(struc ~ env)
OLS # print regression results to the screen
abline(OLS, col="red4")
plot(dist, struc, xlab="Geographic Distance", ylab="1 - Bray-Curtis", main="Community\nComposition", col='darkorchid4')
OLS = lm(struc ~ dist)
abline(OLS, col="red4")

# 10) Use `simba` to calculates the difference in slope or intercept of two regression lines
diffslope(env, struc, dist, struc)
```

***Question 7***: What can you conclude about community similarity with regard to environmental distance and geographic distance?

> ***Answer 7***: Community similarity decreases with environmenal distance and does not differ with respect to geographic distance- near environments have greater similarity than far ones. communities close to one another are more similar than communities that are farther away. 


## 7) EXAMINING SPECIES SPATIAL ABUNDANCE DISTRIBUTIONS

***Question 8***: In your own words, explain the species spatial abundance distribution and what it reveals.

> ***Answer 8***: Spatial abundance distributions reveal the frequency at which we find individuals at particular abundances - patterns of spatial aggregation (distribution of individuals within a landscape)


In the R code chunk below, do the following: 

1. Define a function that will generate the SSAD for a given OTU.
2. Draw six OTUs at random from the IN ponds dataset and and plot their SSADs as kernel density curves. Use **while loops** and **if** statements to accomplish this.

```{r}
# 1. Define an SSAD function
ssad = function(x){
  ad = c(2, 2)
  ad = OTUs[, otu]
  ad = as.vector(t(x = ad))
  ad = ad[ad > 0]
}
# 2. Set plot parameters
par(mfrow=c(2, 2))
# 3. Declare a counter variable
ct = 0 
# 4. Write a while loop to plot the SSADs of six species chosen at random 
while (ct < 4){
otu = sample(1:length(OTUs), 1) 
ad = ssad(otu) 
if (length(ad) > 10 & sum(ad > 100)){ 
ct = ct + 1
plot(density(ad), col = 'red', xlab='Site abundance',
ylab='Probability Density', main = otu.names[otu])
} 
}
```

## 8) UNDERSTANDING SPATIAL SCALE
Many patterns of biodiversity relate to spatial scale.

***Question 9***:  List, describe, and give examples of the two main aspects of spatial scale

> ***Answer 9***: 


## 9) CONSTRUCTING THE SPECIES-AREA RELATIONSHIP

***Question 10***: In your own words, describe the species-area relationship.

> ***Answer 10***: 


In the R code chunk below, provide the code to: 

1. Simulate the spatial distribution of a community with 100 species, letting each species have between 1 and 1,000 individuals.

```{r, results = 'hide', message = FALSE, warning = FALSE}
# 1. Declare variables to hold simulated community and species information 

# 2. Populate the simulated landscape


```

While consult the handout for assistance, in the R chunk below, provide the code to: 

1. Use a nested design to examine the SAR of our simulated community.
2. Plot the SAR and regression line. 

```{r}
# 1. Declare the spatial extent and lists to hold species richness and area data

# 2. Construct a 'while' loop and 'for' loop combination to quantify the numbers of species for progressively larger areas of the simulated landscape.

# 3. Be sure to log10-transform the richness and area data
```


In the R code chunk below, provide the code to: 

1. Plot the richness and area data as a scatter plot.
2. Calculate and plot the regression line
3. Add a legend for the z-value (i.e., slope of the SAR)

```{r, results = 'hide', message = FALSE, warning = FALSE}

```


***Question 10a***: Describe how richness relates to area in our simulated data by interpreting the slope of the SAR.

> ***Answer 10a***: 

***Question 10b***: What does the y-intercept of the SAR represent?

> ***Answer 10b***: 


## SYNTHESIS
Load the dataset you are using for your project. Plot and discuss either the geogrpahic Distance-Decay relationship, the SSADs for at least four species, or any variant of the SAR (e.g., random accumulation of plots or areas, accumulation of contiguous plots or areas, nested design).